encoders/firebase-decoders-json/src/main/java/com/google/firebase/decoders/FieldRef.java

```diff
@@ -0,0 +1,77 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.decoders;
+
+import androidx.annotation.NonNull;
+
+// TODO: support required, optional.
+
+/**
+ * {@link FieldRef} act as an wire to put decoded information to {@link ObjectDecoderContext} and
+ * get decoded information from {@link CreationContext}. It also provide {@link DataDecoder} the
+ * ability to obtain the data type of nextToken.
+ */
+public abstract class FieldRef<T> {
+  private TypeToken<T> typeToken;
+
+  @NonNull
+  public abstract TypeToken<T> getTypeToken();
+
+  /** Used to represent primitive data type. */
+  public static final class Primitive<T> extends FieldRef<T> {
+    @NonNull public static Primitive<Integer> INT = new Primitive<>(TypeToken.of(int.class));
+
+    @NonNull
+    public static Primitive<Boolean> BOOLEAN = new Primitive<>(TypeToken.of(boolean.class));
+
+    @NonNull public static Primitive<Short> SHORT = new Primitive<>(TypeToken.of(short.class));
+
+    @NonNull public static Primitive<Long> LONG = new Primitive<>(TypeToken.of(long.class));
+
+    @NonNull public static Primitive<Float> FLOAT = new Primitive<>(TypeToken.of(float.class));
+
+    @NonNull public static Primitive<Double> DOUBLE = new Primitive<>(TypeToken.of(double.class));
+
+    @NonNull public static Primitive<Character> CHAR = new Primitive<>(TypeToken.of(char.class));
+
+    private Primitive(@NonNull TypeToken<T> typeToken) {
+      super.typeToken = typeToken;
+    }
+
+    @NonNull
+    @Override
+    public TypeToken<T> getTypeToken() {
+      return super.typeToken;
+    }
+  }
+
+  /** Use it to represent Boxed Data type */
+  public static final class Boxed<T> extends FieldRef<T> {
+    @NonNull
+    public static <T> Boxed<T> of(@NonNull TypeToken<T> typeToken) {
+      return new Boxed<T>(typeToken);
+    }
+
+    private Boxed(@NonNull TypeToken<T> typeToken) {
+      super.typeToken = typeToken;
```

Comment:
let's make a precondition check that typeToken is not in fact primitive and throw `IllegalArgumentException` if it is, wdyt?

Reply:
I agree! this way we can ensure `Boxed` only contains non-primitive types and force to create `Primitive` by a pre-defined constant value.

Reply:
👍 can you also add a test for it?